File: Problems/Algorithms/67. Add Binary/add_binary.rs
impl Solution {
    pub fn add_binary(mut a: String, mut b: String) -> String {
        if a.len() < b.len() {
            std::mem::swap(&mut a, &mut b);
        }

        let mut ans = String::new();
        let mut carry = 0;
        let mut chars = b.chars().rev();
        for ch in a.chars().rev() {
            if ch == '1' {
                carry += 1;
            }
            if let Some('1') = chars.next() {
                carry += 1;
            }

            ans.insert(0, if carry & 1 == 1 { '1' } else { '0' });
            carry >>= 1;
        }

        if carry != 0 {
            ans.insert(0, '1');
        }
        ans
    }
}

File: Problems/Algorithms/67. Add Binary/add_binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        i, j = len(a) - 1, len(b) - 1

        results = []
        carry = 0
        while i >= 0 or j >= 0 or carry:
            d1 = int(a[i]) if i >= 0 else 0
            d2 = int(b[j]) if j >= 0 else 0
            results.append(str((d1 + d2 + carry) % 2))
            carry = (d1 + d2 + carry) // 2
            i, j = i-1, j-1

        return ''.join(results[::-1])

